File: setup_gmail_token.py
"""Helper script to get Gmail OAuth token for Netlify deployment"""
import os
import json
import base64
from gmail_client import GmailClient, SCOPES

def main():
    print("=" * 60)
    print("Gmail Token Setup for Netlify")
    print("=" * 60)
    print()
    
    # Check for credentials file
    credentials_file = "credentials.json"
    if not os.path.exists(credentials_file):
        print(f"ERROR: {credentials_file} not found!")
        print("Please download credentials.json from Google Cloud Console")
        print("and place it in this directory.")
        return
    
    print(f"Found {credentials_file}")
    print("Starting OAuth flow...")
    print("A browser window will open for authentication.")
    print()
    
    # Create Gmail client and authenticate
    gmail = GmailClient(credentials_file, "token.json")
    
    try:
        gmail.authenticate()
        print("✓ Authentication successful!")
        print()
        
        # Read the token file
        with open("token.json", "r") as f:
            token_data = json.load(f)
        
        # Display token info
        print("Token saved to token.json")
        print()
        print("=" * 60)
        print("For Netlify deployment, add this to GMAIL_TOKEN_JSON:")
        print("=" * 60)
        print()
        
        # Option 1: JSON string
        token_json_str = json.dumps(token_data)
        print("Option 1 - Direct JSON (escape quotes in Netlify UI):")
        print(token_json_str[:200] + "..." if len(token_json_str) > 200 else token_json_str)
        print()
        
        # Option 2: Base64 encoded
        token_b64 = base64.b64encode(token_json_str.encode('utf-8')).decode('utf-8')
        print("Option 2 - Base64 encoded (recommended):")
        print(token_b64[:100] + "..." if len(token_b64) > 100 else token_b64)
        print()
        
        print("=" * 60)
        print("Also add GMAIL_CREDENTIALS_JSON to Netlify:")
        print("=" * 60)
        print()
        
        # Read credentials
        with open(credentials_file, "r") as f:
            creds_data = json.load(f)
        
        creds_json_str = json.dumps(creds_data)
        creds_b64 = base64.b64encode(creds_json_str.encode('utf-8')).decode('utf-8')
        
        print("Base64 encoded credentials:")
        print(creds_b64[:100] + "..." if len(creds_b64) > 100 else creds_b64)
        print()
        print("Setup complete! Copy these values to Netlify environment variables.")
        
    except Exception as e:
        print(f"ERROR: {e}")
        print()
        print("Make sure you have:")
        print("1. Downloaded credentials.json from Google Cloud Console")
        print("2. Enabled Gmail API in your Google Cloud project")
        print("3. Created OAuth 2.0 credentials (Desktop app type)")

if __name__ == "__main__":
    main()

File: setup_gmail_token.js
/** Script to refresh Gmail OAuth token */
const { OAuth2Client } = require('google-auth-library');
const fs = require('fs').promises;
const path = require('path');
const readline = require('readline');

// Gmail API scopes
const SCOPES = [
  'https://www.googleapis.com/auth/gmail.readonly',
  'https://www.googleapis.com/auth/gmail.send'
];

const TOKEN_PATH = path.join(__dirname, 'token.json');
const CREDENTIALS_PATH = path.join(__dirname, 'credentials.json');

/**
 * Reads previously authorized token from a file, or creates a new one.
 */
async function getNewToken(oAuth2Client) {
  const authUrl = oAuth2Client.generateAuthUrl({
    access_type: 'offline',
    scope: SCOPES,
  });
  
  console.log('Authorize this app by visiting this url:', authUrl);
  
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  
  return new Promise((resolve, reject) => {
    rl.question('Enter the code from that page here: ', async (code) => {
      rl.close();
      try {
        const { tokens } = await oAuth2Client.getToken(code);
        oAuth2Client.setCredentials(tokens);
        
        // Save the token for future program executions
        await fs.writeFile(TOKEN_PATH, JSON.stringify(tokens));
        console.log('Token stored to', TOKEN_PATH);
        
        // Also output as base64 for easy copying to Netlify
        const tokenBase64 = Buffer.from(JSON.stringify(tokens)).toString('base64');
        console.log('\n=== Token (Base64 encoded for Netlify) ===');
        console.log(tokenBase64);
        console.log('\n=== Token (JSON - for direct copy) ===');
        console.log(JSON.stringify(tokens));
        console.log('\nTo set in Netlify, run:');
        console.log(`netlify env:set GMAIL_TOKEN_JSON "${JSON.stringify(tokens)}"`);
        console.log('Or use the base64 version:');
        console.log(`netlify env:set GMAIL_TOKEN_JSON "${tokenBase64}"`);
        
        resolve(tokens);
      } catch (err) {
        reject(err);
      }
    });
  });
}

/**
 * Load or request authorization to call APIs.
 */
async function authorize() {
  let credentials;
  
  try {
    // Check for credentials in environment variable first
    if (process.env.GMAIL_CREDENTIALS_JSON) {
      try {
        credentials = JSON.parse(Buffer.from(process.env.GMAIL_CREDENTIALS_JSON, 'base64').toString('utf-8'));
      } catch (e) {
        credentials = JSON.parse(process.env.GMAIL_CREDENTIALS_JSON);
      }
    } else {
      // Load from file
      const content = await fs.readFile(CREDENTIALS_PATH);
      credentials = JSON.parse(content);
    }
  } catch (err) {
    console.error('Error loading client secret file:', err);
    console.error('\nPlease ensure credentials.json exists or GMAIL_CREDENTIALS_JSON is set.');
    process.exit(1);
  }
  
  const { client_secret, client_id, redirect_uris } = credentials.installed || credentials.web;
  const oAuth2Client = new OAuth2Client(
    client_id,
    client_secret,
    redirect_uris[0]
  );
  
  // Check if we have previously stored a token.
  let token;
  try {
    const content = await fs.readFile(TOKEN_PATH);
    token = JSON.parse(content);
  } catch (err) {
    // No token file, will get new one
  }
  
  // Check command line argument to force new token
  const forceNew = process.argv.includes('--force') || process.argv.includes('-f');
  
  if (token && !forceNew) {
    oAuth2Client.setCredentials(token);
    
    // Try to refresh if expired
    if (token.expiry_date && token.expiry_date < Date.now()) {
      console.log('Token expired, refreshing...');
      try {
        const { credentials: newToken } = await oAuth2Client.refreshAccessToken();
        oAuth2Client.setCredentials(newToken);
        
        // Save refreshed token
        await fs.writeFile(TOKEN_PATH, JSON.stringify(newToken));
        console.log('Token refreshed and saved!');
        
        // Output for Netlify
        const tokenBase64 = Buffer.from(JSON.stringify(newToken)).toString('base64');
        console.log('\n=== Refreshed Token (Base64) ===');
        console.log(tokenBase64);
        console.log('\n=== Refreshed Token (JSON) ===');
        console.log(JSON.stringify(newToken));
        console.log('\nTo set in Netlify:');
        console.log(`netlify env:set GMAIL_TOKEN_JSON "${JSON.stringify(newToken)}"`);
        
        return newToken;
      } catch (err) {
        console.log('Could not refresh token, getting new one...');
        console.error('Error:', err.message);
        // Fall through to get new token
      }
    } else {
      console.log('Token is still valid!');
      console.log('To force a new token, run: npm run setup-gmail -- --force');
      console.log('\nCurrent token (for Netlify):');
      const tokenBase64 = Buffer.from(JSON.stringify(token)).toString('base64');
      console.log('Base64:', tokenBase64);
      console.log('JSON:', JSON.stringify(token));
      return token;
    }
  }
  
  // Get new token
  console.log('Getting new token...');
  return await getNewToken(oAuth2Client);
}

// Run the authorization
authorize()
  .then((token) => {
    console.log('\n✓ Token setup complete!');
    process.exit(0);
  })
  .catch((err) => {
    console.error('Error:', err);
    process.exit(1);
  });

